server/lib/realtime/workflows.ex

```diff
@@ -0,0 +1,238 @@
+  """
+  def delete_workflow(workflow) do
+    Repo.delete(workflow)
+  end
+
+  ## Execution
+
+  @doc """
+  Creates a new workflow execution, using the most recent workflow revision.
+  """
+  def create_workflow_execution(workflow_id, attrs \\ %{}) do
+    with {:ok, revision} <- get_latest_workflow_revision(workflow_id),
+         {:ok, execution} <- insert_execution_with_revision(revision, attrs) do
+      {:ok, %{execution: execution, revision: revision}}
+    end
+  end
+
+  @doc """
+  Returns the workflow execution with the given id.
+  """
+  def get_workflow_execution(id) do
+    Execution
+    |> Repo.get(id)
+    |> Repo.preload([:revision])
+    |> value_or_not_found(id)
+  end
+
+  @doc """
+  Returns a list of executions for the given workflow.
+  """
+  def list_workflow_executions(workflow_id) do
+    from(rev in Revision, where: [workflow_id: ^workflow_id], preload: :executions)
+    |> Repo.all()
+    |> Enum.flat_map(fn rev ->
+      # Return a list of %{execution, revision} maps like other functions so it's convenient to render
+      Enum.map(rev.executions, fn exec -> %{execution: exec, revision: rev} end)
+    end)
+  end
+
+  @doc """
+  Deletes the given workflow execution.
+  """
+  def delete_workflow_execution(execution) do
+    Repo.delete(execution)
+  end
+
+  @doc """
+  Creates a new workflow execution and waits for the workflow to finish.
+
+  ## Options
+
+  * `:timeout` - How long to wait for the workflow to finish.
+  """
+  def invoke_workflow_and_wait_for_reply(workflow, attrs, opts \\ []) do
+    timeout = Keyword.get(opts, :timeout, 10_000)
+
+    with {:ok, %{execution: execution, revision: revision}} <-
+           create_workflow_execution(workflow.id, attrs) do
+      # Start and await a Task so the original self() can wait for other messages.
+      # In practice this is needed to test the ExecutionController.
+      task =
+        Task.async(fn ->
+          with {:ok, _pid} <-
+                 Interpreter.start_transient(workflow, execution, revision, reply_to: self()) do
+            receive do
+              {:succeed, msg} -> {:ok, msg, execution, revision}
+              {:fail, msg} -> {:error, msg, execution, revision}
+              err -> {:error, err, execution, revision}
+            after
+              timeout -> {:timeout, execution, revision}
+            end
+          end
+        end)
+
+      Task.await(task, timeout)
+    end
+  end
+
+
+  @doc """
+  Create and start the execution of all workflows that are triggered by `txn`.
+  """
+  def invoke_transaction_workflows(%Transaction{changes: [_ | _]} = txn) do
+    workflows = Manager.workflows_for_change(txn)
+
+    txn_as_map = Map.from_struct(txn)
+
+    attrs = %{
+      arguments: txn_as_map,
+    }
+
+    Enum.each(workflows, fn workflow ->
+      with {:ok, %{execution: execution, revision: revision}} <- create_workflow_execution(workflow.id, attrs) do
+	# TODO: should be based on default execution type
```

Comment:
Todo: we should which type of workflow this is (it's always running as a persistent workflow right now